packages/snaps-utils/src/vfile/vfile.ts

```diff
@@ -0,0 +1,103 @@
+// TODO(ritave): Move into separate package @metamask/vfile / @metamask/utils + @metamask/to-vfile when passes code review
+// TODO(ritave): Streaming vfile contents similar to vinyl maybe?
+// TODO(ritave): Move fixing manifest to write messages to vfile similar to unified instead of throwing "ProgrammaticallyFixableErrors".
+//               Better yet, move manifest fixing into eslint fixing altogether
+import { assert, hasProperty } from '@metamask/utils';
+import { instance, integer, is, size, type } from 'superstruct';
+
+// Using https://github.com/vfile/vfile would be helpful, but they only support ESM.
+// https://github.com/gulpjs/vinyl is also good, but they normalize paths, which we can't do, because
+// we're calculating checksums based on original path.
+
+/**
+ * This map registers the type of the `data` key of a `VFile`.
+ *
+ * This type can be augmented to register custom `data` types.
+ *
+ * @example
+ * declare module '@metamask/vfile' {
+ *   interface DataMap {
+ *     // `file.data.name` is typed as `string`
+ *     name: string
+ *   }
+ * }
+ */
```

Comment:
What's the use-case of this? Why couldn't we do this using generics?

Reply:
The use case is allowing modules to add their own metadata. In our example, it's [SnapLocations adding canonical paths](https://github.com/MetaMask/snaps-monorepo/pull/1019/files#diff-113f19d275e0849e32e8d55edf2a55c2772c744149099c9505e657fbb3fdaed4R8).

If vfile are to be decoupled from Snaps and be put in `@metamask/utils`, different modules and libraries will put different things there, and this mechanism allows vfile types to be extendable, generics don't solve extensibility.